### PR TITLE
Fix outfitted track speed ignoring track type

### PR DESCRIPTION
**Pull request by liach**
This allows track kits to determine the track type in order to handle the speed correctly. Added some reasonable annotations, too.

### CODE REVIEW

File: mods/railcraft/api/tracks/ITrackKitInstance.java [hunk 1]
```diff
@@ -125,8 +125,8 @@ default World theWorld() {
      * @param cart The cart on the rail, may be null.
      * @return The max speed of the current rail.
      */
-    default float getRailMaxSpeed(World world, EntityMinecart cart, BlockPos pos) {
-        return 0.4F;
+    default float getRailMaxSpeed(World world, @Nullable EntityMinecart cart, BlockPos pos, TrackType type) {
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/ITrackKitInstance.java]: TrackType is totally not needed as a parameter.

File: mods/railcraft/api/tracks/ITrackKitInstance.java [hunk 2]
```diff
@@ -136,4 +136,11 @@ default boolean isProtected() {
         return false;
     }
 
+    /**
+     * Returns the track type of this track.
+     */
+    default TrackType getTrackType() {
+        return TrackToolsAPI.getTypeForTrackKitInstance(this);
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/ITrackKitInstance.java]:
```
return ((IOutfittedTrackTile)getTile()).getTrackType();
```

File: mods/railcraft/api/tracks/ITrackKitInstance.java [hunk 3]
```diff
@@ -125,8 +125,8 @@ default World theWorld() {
      * @param cart The cart on the rail, may be null.
      * @return The max speed of the current rail.
      */
-    default float getRailMaxSpeed(World world, EntityMinecart cart, BlockPos pos) {
-        return 0.4F;
+    default float getRailMaxSpeed(World world, @Nullable EntityMinecart cart, BlockPos pos, TrackType type) {
+        return type.getEventHandler().getMaxSpeed(world, cart, pos);
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/ITrackKitInstance.java]:
```
return getTrackType().getEventHandler().getMaxSpeed(world, cart, pos);
```

File: mods/railcraft/api/tracks/TrackToolsAPI.java [hunk 4]
```diff
@@ -201,4 +206,21 @@ public static boolean isTrackFuzzyAt(World world, BlockPos pos) {
...
+
+    private static Function<TileEntity, TrackType> trackTypeFunction;
+
+    /**
+     * Other mods don't use this!
+     */
+    @Deprecated
+    public static void setTrackTypeFunction(Function<TileEntity, TrackType> function) {
+        trackTypeFunction = function;
+    }
```
Comment (CovertJaguar) [inline mods/railcraft/api/tracks/TrackToolsAPI.java]: None of this is needed.

**Review by CovertJaguar — CHANGES_REQUESTED** [review]

File: mods/railcraft/api/tracks/ITrackKitInstance.java (hunk 2 above)
Comment (liach) [reply mods/railcraft/api/tracks/ITrackKitInstance.java]: The dummy tile will crash here

Comment (CovertJaguar) [reply mods/railcraft/api/tracks/ITrackKitInstance.java]: Why would it do that? Its designed to return the fallback TrackType.

**Comment by liach** [comment]
Will fix them when I'm home.

**Review by CovertJaguar — APPROVED** [review]

**Pull request merged**